### DISCUSSION TITLE
Can I put some array as a field in a java class that implements iterator, then iterating over it using next?

### DISCUSSION THREAD

**Discussion by asmmo** [jni-rs/jni-rs#794] [Q&A]
Can I put some array as a field in a java class that implements java `iterator`, then iterating over it using `next()`. and will the java `GC` deallocate it after the object of that class goes out of scope?
The objects of the mentioned class will be constructed in the rust-side but `next` method will call some native method which will retrieve the next element in the array.

**Comment by maurolacy** [comment]
Why not? If you do everything correctly, that should work.

In general, and please correct me if I'm wrong, the only memory that will not be deallocated or GC'd is the one in the native (Rust) heap, that you explicitly mark as manually managed (through `into_raw` and such).

**Comment by rib** [comment]
This is a pretty old question but essentially I think the answer is 'yes' as suggested above.

The `jni` crate will generally reference objects with "local references" that are temporary references that are only associated with the current JNI stack frame.

Local references won't keep objects alive indefinitely (stopping them from being garbage collected) and so there wouldn't typically be any reason why Java objects constructed via Rust/JNI would not get garbage collected the same as they would have done if allocated in Java.

Some exceptions to this would be:
- If you explicitly create a global reference for those objects (`env.new_global_ref()`) that you then leak so the global reference is never deleted. These leaked references could permanently keep your objects alive (that wouldn't have to be done quite intentionally)
- If you have a long-running thread with a JNI stack frame holding local references to these objects which never unwinds to release those local references. If this were happening then you effectively also have a leak and could eventually run out of memory / reach the limit on how many local references you can allocate. 

Regarding the last point: If your objects are allocated from a long-running thread it would be important to ensure that `env.with_local_frame()` would be called by something before allocating your new object and adding to your array so then when you return from `with_local_frame` your code would pop the current JNI stack frame to release any temporary local reference to those new array objects. Alternatively you could also explicitly delete the local reference immediately after adding to the array (via `env.delete_local_ref()`).

Note that `vm.attach_current_thread()` will internally also call `with_local_frame()` and since that would typically be called as a wrapper around code using JNI from a long-running thread then it should normally Just Work(tm) to avoid the above kind of leak.
